services/alert/handlers.go

```diff
@@ -506,54 +506,84 @@ func (h *matchHandler) Handle(event alert.Event) {
 	}
 }
 
+var changedFuncSignature = map[stateful.Domain]ast.ValueType{}
+var levelFuncSignature = map[stateful.Domain]ast.ValueType{}
+var nameFuncSignature = map[stateful.Domain]ast.ValueType{}
+var taskNameFuncSignature = map[stateful.Domain]ast.ValueType{}
+var durationFuncSignature = map[stateful.Domain]ast.ValueType{}
+
+func init() {
+	d := stateful.Domain{}
+	changedFuncSignature[d] = ast.TBool
+	levelFuncSignature[d] = ast.TInt
+	nameFuncSignature[d] = ast.TString
+	taskNameFuncSignature[d] = ast.TString
+	durationFuncSignature[d] = ast.TDuration
+}
+
 func (h *matchHandler) match(event alert.Event) (bool, error) {
 	// Populate scope
 	h.scope.Reset()
 
 	h.logger.Printf("D! match %+v", h)
 
 	if h.usesChanged {
-		h.scope.SetDynamicMethod(changedFunc, func(self interface{}, args ...interface{}) (interface{}, error) {
-			if len(args) != 0 {
-				return nil, fmt.Errorf("%s takes no arguments", changedFunc)
-			}
-			return event.State.Level != event.PreviousState().Level, nil
+		h.scope.SetDynamicFunc(changedFunc, &stateful.DynamicFunc{
+			F: func(self interface{}, args ...interface{}) (interface{}, error) {
+				if len(args) != 0 {
```

Comment:
While it would be a violation of the signature, there's nothing that prevents a user from deciding to call `F` themselves without first checking the signature. Especially since `F` is exported. Thoughts?